zcash_client_sqlite/src/wallet/init/migrations/fix_broken_commitment_trees.rs

```diff
@@ -0,0 +1,82 @@
+//! Truncates away bad note commitment tree state for users whose wallets were broken by incorrect
+//! reorg handling.
+use std::collections::HashSet;
+
+use rusqlite::{self, OptionalExtension};
+use schemer;
+use schemer_rusqlite::RusqliteMigration;
+use uuid::Uuid;
+use zcash_protocol::consensus::{self, BlockHeight};
+
+use crate::wallet::{
+    self,
+    init::{migrations::support_legacy_sqlite, WalletMigrationError},
+};
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x9fa43ce0_a387_45d1_be03_57a3edc76d01);
+
+const DEPENDENCIES: &[Uuid] = &[support_legacy_sqlite::MIGRATION_ID];
+
+pub(super) struct Migration<P> {
+    pub(super) params: P,
+}
+
+impl<P> schemer::Migration for Migration<P> {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        DEPENDENCIES.iter().copied().collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Truncates away bad note commitment tree state for users whose wallets were broken by bad reorg handling."
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        #[cfg(not(feature = "orchard"))]
+        let max_height_query = r#"
+            SELECT MAX(height) FROM blocks
+            JOIN sapling_tree_checkpoints sc ON sc.checkpoint_id = height
+        "#;
+        #[cfg(feature = "orchard")]
+        let max_height_query = r#"
+            SELECT MAX(height) FROM blocks
+            JOIN sapling_tree_checkpoints sc ON sc.checkpoint_id = height
+            JOIN orchard_tree_checkpoints oc ON oc.checkpoint_id = height
+        "#;
+
+        let max_block_height = transaction
+            .query_row(max_height_query, [], |row| {
+                let cid = row.get::<_, Option<u32>>(0)?;
+                Ok(cid.map(BlockHeight::from))
+            })
+            .optional()?
+            .flatten();
+
+        if let Some(h) = max_block_height {
+            wallet::truncate_to_height(transaction, &self.params, h)?;
```

Comment:
For wallets that did not have this problem, the effect here should just be that any call their wallet made to `update_chain_tip` or adding subtree roots, gets reverted?

Reply:
Correct.